src/Dictionaries/HierarchyDictionariesUtils.h

```diff
@@ -0,0 +1,468 @@
+            hierarchy_null_value,
+            std::forward<IsKeyValidFunc>(is_key_valid_func),
+            std::forward<GetParentKeyFunc>(get_parent_func));
+
+        auto & offsets = hierarchy.offsets;
+        auto & elements = hierarchy.elements;
+
+        for (size_t i = 0; i < offsets.size(); ++i)
+        {
+            size_t i_elements_start = i > 0 ? offsets[i - 1] : 0;
+            size_t i_elements_end = offsets[i];
+
+            auto & key_to_find = in_keys[i];
+
+            const auto * begin = elements.begin() + i_elements_start;
+            const auto * end = elements.begin() + i_elements_end;
+
+            const auto * it = std::find(begin, end, key_to_find);
+
+            bool contains_key = (it != end);
+            result[i] = contains_key;
+        }
+
+        return result;
+    }
+
+    struct GetAllDescendantsStrategy { size_t level = 0; };
+    struct GetDescendantsAtSpecificLevelStrategy { size_t level = 0; };
+
+    /** Get descendants for keys iterating the hierarchy from parent to child using parent_to_child hash map provided by client.
+      * GetAllDescendantsStrategy get all descendants for key
+      * GetDescendantsAtSpecificLevelStrategy get descendants only for specific hierarchy level.
+      * Hierarchy result is ElementsAndOffsets structure, for each element there is descendants array,
+      * with size offset[element_index] - (element_index > 0 ? offset[element_index - 1] : 0).
+      *
+      * Example:
+      * id  parent_id
+      * 1   0
+      * 2   1
+      * 3   1
+      * 4   2
+      *
+      * Example. Strategy GetAllDescendantsStrategy.
+      * Requested keys [0, 1, 2, 3, 4].
+      * Result: [1, 2, 3, 4], [2, 2, 4], [4], [], []
+      * Elements: [1, 2, 3, 4, 2, 3, 4, 4]
+      * Offsets: [4, 7, 8, 8, 8]
+      *
+      * Example. Strategy GetDescendantsAtSpecificLevelStrategy with level 1.
+      * Requested keys [0, 1, 2, 3, 4].
+      * Result: [1], [2, 3], [4], [], [];
+      * Offsets: [1, 3, 4, 4, 4];
+      */
+    template <typename KeyType, typename Strategy>
+    ElementsAndOffsets<KeyType> getDescendants(
+        const PaddedPODArray<KeyType> & keys,
+        const HashMap<KeyType, PaddedPODArray<KeyType>> & parent_to_child,
+        Strategy strategy)
+    {
+        /// If strategy is GetAllDescendantsStrategy we try to cache and later reuse previously calculated descendants.
+        /// If strategy is GetDescendantsAtSpecificLevelStrategy we does not use cache strategy.
+        size_t keys_size = keys.size();
+
+        PaddedPODArray<KeyType> descendants;
+        descendants.reserve(keys_size);
+
+        PaddedPODArray<size_t> descendants_offsets;
+        descendants_offsets.reserve(keys_size);
+
+        struct Range
+        {
+            size_t start_index;
+            size_t end_index;
+        };
+
+        static constexpr Int64 key_range_requires_update = -1;
+        HashMap<KeyType, Range> already_processed_keys_to_range [[maybe_unused]];
```

Comment:
Maybe better to use HashMap<KeyType, std::set<KeyType>> ? The code will be simplier

Reply:
We cannot use HashMap here, because we will need to copy sets.